### PR TITLE
[2804] - Centralise geocoding logic

### PR DESCRIPTION
**Pull request by TobyRet**
### Context
MCB's geocoding logic did not restrict searches to the UK region. This meant that geocoding of any invalid Sites (specifically those without an address) would result in coordinates set to somewhere in the U.S (this is the default behaviour).

### Changes proposed in this pull request
- Extract geocoding logic to a new service. Geocoding is restricted to the UK (gb).
- Update mcb console and background job to use this new service

Centralising this logic will (hopefully) reduce the chances of bugs resulting from diverging paths in future 

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by TobyRet** [comment]
[MCB geocode commands need to be restricted to UK](https://trello.com/c/rUbLtL6K/2804-mcb-geocode-commands-need-to-be-restricted-to-uk)

**Comment by TobyRet** [comment]
Cache will need to be expired before re-geocoding
https://github.com/alexreisner/geocoder#caching

File: lib/mcb.rb [hunk 1]
```diff
@@ -478,12 +478,22 @@ def launch_repl?(argv)
...
-              "New lat/long #{obj.latitude},#{obj.latitude} for full_address '#{obj.full_address}'"
+      results = Geocoder.search(obj.full_address, params: { region: "gb" })
+
+      if results.present?
+        result = results.first
+        obj.latitude = result.latitude
+        obj.longitude = result.longitude
+        saved = obj.save(validate: !force)
+
+        warn "Updating failed for #{obj.class}:#{obj.id} - #{obj}. Error: #{obj.errors}" unless saved
```
Comment (timabell) [inline lib/mcb.rb]: I think move this into an `else` for the `if` below

File: lib/mcb.rb [hunk 2]
```diff
@@ -478,12 +478,22 @@ def launch_repl?(argv)
...
+        if saved
+          verbose "Geocoded #{obj.class}:#{obj.id} - #{obj}. " \
+                "New lat/long #{obj.latitude},#{obj.latitude} for full_address '#{obj.full_address}'"
```
Comment (timabell) [inline lib/mcb.rb]:
```suggestion
                "New lat/long #{obj.latitude},#{obj.longitude} for full_address '#{obj.full_address}'"
```

File: spec/services/geocoder_service_spec.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          to raise_error(ActiveRecord::RecordInvalid)
+    end
+
+    it "geocodes an invalid object if forced" do
+      expect { GeocoderService.geocode(obj: invalid_site, force: true) }.
+          to change { invalid_site.reload.latitude }.from(nil).to(51.4524877).
+              and change { invalid_site.longitude }.from(nil).to(-0.1204749)
+    end
+
+    it "geocodes UK (gb) addresses only" do
```
Comment (TobyRet) [inline spec/services/geocoder_service_spec.rb]: @defong - I hope you sleep easy tonight 😄

File: app/models/site.rb [hunk 4]
```diff
@@ -63,7 +63,7 @@ def needs_geolocation?
   def full_address
     address = [address1, address2, address3, address4, postcode]
 
-    unless location_name.downcase == MAIN_SITE
+    unless location_name && location_name.downcase == MAIN_SITE
```
Comment (defong) [inline app/models/site.rb]: why was this changed?

File: app/services/geocoder_service.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class GeocoderService
+  def self.geocode(obj:, force: false)
+    results = Geocoder.search(obj.full_address, params: { region: "gb" })
+
+    if results.present?
+      result = results.first
+      obj.latitude = result.latitude
+      obj.longitude = result.longitude
+      obj.save(validate: !force)
```
Comment (defong) [inline app/services/geocoder_service.rb]: should it be the service that save its?

File: spec/services/geocoder_service_spec.rb (hunk 3 above)
Comment (defong) [reply spec/services/geocoder_service_spec.rb]: just wanna make sure you snooze like a cuddly :baby:...

File: app/models/site.rb (hunk 4 above)
Comment (TobyRet) [reply app/models/site.rb]: Sorry this was an error. Changing back
